demos/coreMQTT_Agent/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,560 @@
+    }
+
+    LogInfo( ( "Received incoming publish message %s", cTerminatedString ) );
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvSubscribeToTopic( MQTTQoS_t xQoS,
+                                 char * pcTopicFilter )
+{
+    MQTTStatus_t xCommandAdded;
+    BaseType_t xCommandAcknowledged = pdFALSE;
+    uint32_t ulSubscribeMessageID;
+    MQTTAgentSubscribeArgs_t xSubscribeArgs;
+    MQTTSubscribeInfo_t xSubscribeInfo;
+    static int32_t ulNextSubscribeMessageID = 0;
+    MQTTAgentCommandContext_t xApplicationDefinedContext = { 0 };
+    MQTTAgentCommandInfo_t xCommandParams = { 0 };
+
+    /* Create a unique number of the subscribe that is about to be sent.  The number
+     * is used as the command context and is sent back to this task as a notification
+     * in the callback that executed upon receipt of the subscription acknowledgment.
+     * That way this task can match an acknowledgment to a subscription. */
+    xTaskNotifyStateClear( NULL );
+    taskENTER_CRITICAL();
+    {
+        ulNextSubscribeMessageID++;
+        ulSubscribeMessageID = ulNextSubscribeMessageID;
+    }
+    taskEXIT_CRITICAL();
+
+    /* Complete the subscribe information.  The topic string must persist for
+     * duration of subscription! */
+    xSubscribeInfo.pTopicFilter = pcTopicFilter;
+    xSubscribeInfo.topicFilterLength = ( uint16_t ) strlen( pcTopicFilter );
+    xSubscribeInfo.qos = xQoS;
+    xSubscribeArgs.pSubscribeInfo = &xSubscribeInfo;
+    xSubscribeArgs.numSubscriptions = 1;
+
+    /* Complete an application defined context associated with this subscribe message.
+     * This gets updated in the callback function so the variable must persist until
+     * the callback executes. */
+    xApplicationDefinedContext.ulNotificationValue = ulNextSubscribeMessageID;
+    xApplicationDefinedContext.xTaskToNotify = xTaskGetCurrentTaskHandle();
+    xApplicationDefinedContext.pArgs = ( void * ) &xSubscribeArgs;
+
+    xCommandParams.blockTimeMs = mqttexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = prvSubscribeCommandCallback;
+    xCommandParams.pCmdCompleteCallbackContext = ( void * ) &xApplicationDefinedContext;
+
+    /* Loop in case the queue used to communicate with the MQTT agent is full and
+     * attempts to post to it time out.  The queue will not become full if the
+     * priority of the MQTT agent task is higher than the priority of the task
+     * calling this function. */
+    LogInfo( ( "Sending subscribe request to agent for topic filter: %s with id %d",
+               pcTopicFilter,
+               ( int ) ulSubscribeMessageID ) );
+
+    do
+    {
+        /* TODO: prvIncomingPublish as publish callback. */
+        xCommandAdded = MQTTAgent_Subscribe( &xGlobalMqttAgentContext,
+                                             &xSubscribeArgs,
+                                             &xCommandParams );
+    } while( xCommandAdded != MQTTSuccess );
```

Comment:
Why is `MQTTAgent_Subscribe` call in a loop? Does this mean that enqueuing of the subsription command will be blocking until it succeeds? 
Can this cause an overfilling of the agent queue in the failure case that a broker has suddenly severed an MQTT connection, and the re-connection attempts in the main task are failing but in the meanwhile multiple demo tasks are stuck in adding subscription commands because the queue is full?

Reply:
`MQTTAgent_Subscribe` will return success if the command is added to the queue. This will keep retrying until the command is added once. Yes, it's possible for the queue to become full here in the case when both the connection has failed and there are more tasks than the size of the queue.

The queue shouldn't really ever become full when the agent task is running since it has a higher priority then the subscribe-publish tasks.

Reply:
What's the value of having blocking nature for the subscribe operation then if it leaves a door open for the task to be permanently blocked if the size of the queue is smaller than the number of tasks..
Should a single large blocking time be used instead of blocking for ever, and treat the unsuccessful queueing of operation as a demo error?

Reply:
I'll just remove the loop in that case